[torch] Expose dynamic_shapes api at multiple levels

BoyuanFeng opened the pull request:
Summary: Exposes `dynamic_shapes` api at multiple levels so it's easier to replace the old API `dynamic_dim()` with the new API `Dim()`.

Test Plan: CI

Differential Revision: D53246409




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -512,16 +514,48 @@ def _export(
-    constraints = constraints or []
+    from .dynamic_shapes import _process_dynamic_shapes
+
+    if constraints is not None:
```
If you replaced all the callsites where we used constraints, why do we need to warn it here? Can't we just error or delete this kwarg directly?

tugsbayasgalan commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1209,6 +1211,9 @@ def export(
     _assume_static_by_default = assume_static_by_default
 
     def inner(*args, **kwargs):
+        nonlocal constraints, f
```
nit: is there a way to do this without referencing non local variables?

ydwu4 reviewed (approved):
LGTM! Left some minor comments in associated diff.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -512,16 +514,48 @@ def _export(
+         range of shapes. See :func:`dynamic_dim` docstring for examples on
+         how to use it.
+
+        dynamic_shapes: Should either be:
```
Formatting seems bit weird? Maybe delete "Should either be"?

BoyuanFeng committed:
[torch] Expose dynamic_shapes api at multiple levels (#118695)

Summary:

Exposes `dynamic_shapes` api at multiple levels so it's easier to replace the old API `dynamic_dim()` with the new API `Dim()`.

Test Plan: CI

Reviewed By: ydwu4

Differential Revision: D53246409
